subsys/net/ip/net_context.c

```diff
@@ -816,6 +816,24 @@ NET_CONN_CB(tcp_established)
 				     sys_get_be32(NET_TCP_HDR(pkt)->ack));
 	}
 
+//printk("%u %u\n", sys_get_be32(NET_TCP_HDR(pkt)->seq), context->tcp->send_ack);
+
+	/* Peer sent us packet we've already seen. Apparently, our ack was lost. */
+	if (sys_get_be32(NET_TCP_HDR(pkt)->seq) < context->tcp->send_ack) {
+printk("Received already seen pkt - ACKing\n");
+
+	set_appdata_values(pkt, IPPROTO_TCP);
```

Comment:
So (sorry, late to the game)... I think I understand the point here.  We don't retransmit an ACK if we already sent it once, and that's wrong.

Nonetheless Jukka is exactly right: the ACK must always be for the maximum sequence number we have received (without "holes" of course).  It's legal and expected that the stack will "skip" ACK values, we don't have to ACK every single packet (sort of the whole point of TCP, really).

Seems like the notion in the fix is correct: if we get a old/retransmitted packet we should always synchronously send an ACK (the correct ACK, not a repetition of whatever sequence number we just received!) on the assumption that the original got droped.